services/web/server/src/simcore_service_webserver/computation_api.py

```diff
@@ -218,6 +219,63 @@
     db_engine: Engine, project_id: str, tasks: Dict[str, Dict], replace_pipeline=True
 ):
     # pylint: disable=no-value-for-parameter
+
+    async def _task_already_exists(
+        conn: SAConnection, project_id: str, node_id: str
+    ) -> bool:
+        task_count: int = await conn.scalar(
+            sa.select([sa.func.count()]).where(
+                and_(
+                    comp_tasks.c.project_id == project_id,
+                    comp_tasks.c.node_id == node_id,
+                )
+            )
+        )
+        assert task_count in (
+            0,
+            1,
+        ), f"Uniqueness violated: task_count={task_count}"  # nosec
+        return task_count != 0
+
+    async def _update_task(
+        conn: SAConnection, task: Dict, project_id: str, node_id: str
+    ) -> None:
+        # update task's inputs/outputs
+        io_update = {}
+        task_inputs: str = await conn.scalar(
+            sa.select([comp_tasks.c.inputs]).where(
+                and_(
+                    comp_tasks.c.project_id == project_id,
+                    comp_tasks.c.node_id == node_id,
+                )
+            )
+        )
+        # updates inputs
+        if task_inputs != task["inputs"]:
+            io_update["inputs"] = task["inputs"]
+
+        # update outputs
+        #  NOTE: update ONLY outputs of front-end nodes. The rest are
+        #  updated by backend services (e.g. workers, interactive services)
+        if task["outputs"] and task["node_class"] == NodeClass.FRONTEND:
+            io_update["outputs"] = task["outputs"]
+
+        if io_update:
+            query = (
+                comp_tasks.update()
+                .where(
+                    and_(
+                        comp_tasks.c.project_id == project_id,
+                        comp_tasks.c.node_id == node_id,
+                    )
+                )
+                .values(**io_update)
+            )
+
+            await conn.execute(query)
+
+    # MAIN -----------
+
```

Comment:
Its so much cleaner and better to understand if everything is annotated. I luv it!